### PR TITLE
Encryption of database connection strings in mql database and database sync

### PR DESCRIPTION
**Pull request by rizwan-adam**
Following changes made:

1. Encryption of database connection string while storing in mql database using symmetric encryption
2. Decryption of database connection string while querying database
3. Fernet manager is added to utilities folder
4. Database sync api is added

### CODE REVIEW

File: server/app/api/v1/databases.py [hunk 1]
```diff
@@ -45,13 +46,24 @@ async def connect_to_database(
...
+        engine = create_engine(connection_string)
+        try:
+            engine.connect()
+        except Exception as e:
+            raise HTTPException(
+                status_code=status.HTTP_400_BAD_REQUEST,
+                detail="Unable to connect to the database: " + str(e),
+            )
+        
+        fernet_manager = FernetManager(current_user.hashed_password)
```
Comment (harshshurutech) [inline server/app/api/v1/databases.py:58]: The key should derived using the user's plaintext password, not the hashed password. In order to derive the key from user's password, we can either do it at the login step and save the key as an HTTP cookie for subsequent request or, we can prompt the user to enter their password again when they click on the "Ask Query" button and then derive the key and decrypt the connection string. We can discuss this in detail.

**Review by harshshurutech — CHANGES_REQUESTED** [review]

File: server/app/api/v1/databases.py (hunk 1 above)
Comment (rizwan-adam) [reply server/app/api/v1/databases.py:58]: 1. Approach one of storing in HTTP cookie is highly insecure and never recommended
2. Approach two is acceptable but adds friction during usage and might not be the best option

Instead we could generate a key the fernet manager creates and store it in db but that would be the same as using hashed password.  Ideally the keys should be stored separately in a vault but we do want to complicate the usage of this app for the user

Comment (harshshurutech) [reply server/app/api/v1/databases.py:58]: > * Approach one of storing in HTTP cookie is highly insecure and never recommended

Agree, and that is what I had mentioned in the [doc](https://gist.github.com/harshshurutech/78c8aa2b44d6acdd02ba8b824b2d41fa?permalink_comment_id=5033204#gistcomment-5033204).

> 2. Approach two is acceptable but adds friction during usage and might not be the best option

Yes, it does introduce some friction.

Another approach is to run a local vault service in the Docker container on the same network, which we can use for storing the keys. I am currently looking into it.

Comment (rizwan-adam) [reply server/app/api/v1/databases.py:58]: Or docker works too for now. Let me know what you find

Comment (harshshurutech) [reply server/app/api/v1/databases.py:58]: [Infisical](https://infisical.com/docs/self-hosting/deployment-options/docker-compose) looks like a great open-source solution for secret management. I'll dig into it and keep you posted.

Comment (harshshurutech) [reply server/app/api/v1/databases.py:58]: Another option is [OpenBao](https://openbao.org/)

**Pull request merged**